### PR TITLE
Feat/metadataedit

### LINKED ISSUE #130: Metadata editor
I am currently doing feature development for metadata editing, and some issues need to be discussed.
What UX/UI do you want? Now there are two options
1. Fully use the original interaction
2. Improve some experience

Write down what you think and what you want to improve

### PR DESCRIPTION
**Pull request by Dec-F**
Develop metadata editing features

Edit(@MrTimscampi): Fixes #130

### CODE REVIEW

**Comment by MrTimscampi** [comment]
I really like how this is looking :yum: 

There's some small style changes that would make this look and feel nicer. Here's a quick example:

image

Clearly delimiting the title, tabs and actions makes the dialog look less like a pile of controls.

Structure-wise, I think moving the tabs out of the `form` would be best.

Ideally, the dialog should have a fixed size and not resize to adapt to its content. Only the form part should be scrollable.

I'll test more to give more feedback.

File: pages/metadata/index.vue [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+export default Vue.extend({
+  data() {
+    return {
+      items: [] as ITreeNode[],
+      itemId: ''
+    };
+  },
+  computed: {},
+  async created() {
+    const floders = (await this.$libraryApi.getMediaFolders()).data
```
Comment (MrTimscampi) [inline pages/metadata/index.vue:38]: Typo in the variable name

File: pages/metadata/index.vue [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      .Items as BaseItemDto[];
+
+    this.items = floders.map((dir, index) => {
```
Comment (MrTimscampi) [inline pages/metadata/index.vue:41]: Same type here.

File: pages/metadata/index.vue [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      return {
+        id: dir.Id || index,
+        name: dir.Name,
+        children: [] as ITreeNode[]
+      };
+    });
+  },
+  methods: {
+    async fetchItems(node: ITreeNode) {
+      const userId = this.$auth.user.Id;
```
Comment (MrTimscampi) [inline pages/metadata/index.vue:51]: Avoid unnecessary assignments.

File: pages/metadata/index.vue [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        await this.$userLibraryApi.getItem(
+          { userId, itemId: '' },
+          {
+            query: {
+              ParentId: node.id,
+              SortBy: 'SortName'
+            }
+          }
+        )
+      ).data as unknown) as { Items: BaseItemDto[] }).Items; //
```
Comment (MrTimscampi) [inline pages/metadata/index.vue:62]: This line seems really dirty.

Fixing the `$userLibraryApi` to the new format looks to be fixing the issues you have with types on my end.

Comment (MrTimscampi) [reply pages/metadata/index.vue:62]: As a general rule of thumb, it's best to avoid forcing types on stuff from the API.

The entire purpose of using TypeScript and a generated API client is to warn us of API changes as soon as they hit.

File: plugins/itemUpdateApi.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { Plugin } from '@nuxt/types';
```
Comment (MrTimscampi) [inline plugins/itemUpdateApi.ts:1]: Should be removed to use the new $api instead.

File: plugins/remoteImageApi.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import { Plugin } from '@nuxt/types';
```
Comment (MrTimscampi) [inline plugins/remoteImageApi.ts:1]: Should be removed to use the new $api instead.

**Review by MrTimscampi — CHANGES_REQUESTED** [review]
We use Yarn for this repo, please don't commit `package-lock.json`.

---

First a proper functional review.

Some user stories first:

* I want to edit information in either or both of the General and Details tabs - **Works**
* I want to add a new actor - **Works**
* I want to delete an actor - **Works**
* I want to search for an image - **Works**
* I want to delete an image - **Doesn't work**

Trying to delete an image only outputs a "1" to the console and never deletes the image.

Now for more UX-related issues:

* The dialog resizing is confusing. It should have a fixed size.
* Only the contents on the right should be scrollable, not the entire dialog
* Having the path be an editable field doesn't make sense. Having it be a subtitle below "Edit metadata" would be nicer, but that's not blocking. At least set the field to read only.
* When opening a dialog for an item, closing it, editing the item on another client, then re-opening the same dialog, the information isn't updated
* There is no feedback when saving an item. This leads to the user not being sure if the item was submitted or not. The minimum would be to change the state of the `Save` button to "loading" in order to both prevent clicking multiple times and give the user information about something being done. It should probably handle failure by reverting the state of the Save button and showing an error toast.
* The headings for the forms should not have `:` in them.

**Comment by MrTimscampi** [comment]
In addition to the above reviews, I feel like a lot of the logic in components (Especially around dialogs) should be in Vuex stores.

Generally, the direction we're trying to take with the recent refactorings I've been doing, is to centralize most logic and API calls to Vuex stores, so that they can be triggered from anywhere.

I'm not too sure how you'd organize these and it won't be a blocking issue for this, though, unless you want to refactor now.

**Comment by Dec-F** [comment]
@MrTimscampi 
Thanks for your advice. I have been busy lately, and the subsequent progress may be much slower. I will try to find time to complete the subsequent development

**Comment by MrTimscampi** [comment]
> @MrTimscampi
> Thanks for your advice. I have been busy lately, and the subsequent progress may be much slower. I will try to find time to complete the subsequent development

No worries, there's no rush :)

There's already a lot of work that went into this, so thanks.

**Comment by Dec-F** [comment]
> There is no feedback when saving an item. This leads to the user not being sure if the item was submitted or not. The minimum would be to change the state of the Save button to "loading" in order to both prevent clicking multiple times and give the user information about something being done. It should probably handle failure by reverting the state of the Save button and showing an error toast.

The button should be in the loading state. As for the error message, I think we need to wait for the global dialog system to complete before adding

**Comment by Camc314** [comment]
> > There is no feedback when saving an item. This leads to the user not being sure if the item was submitted or not. The minimum would be to change the state of the Save button to "loading" in order to both prevent clicking multiple times and give the user information about something being done. It should probably handle failure by reverting the state of the Save button and showing an error toast.
> 
> The button should be in the loading state. As for the error message, I think we need to wait for the global dialog system to complete before adding

See 
https://github.com/jellyfin/jellyfin-vue/blob/master/components/LoginForm.vue#L112-L115

File: components/MetadataEditor.vue [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+      const itemInfo = (
+        await this.$userLibraryApi.getItem({
+          userId,
+          itemId: this.itemId
+        })
+      ).data;
+      this.$data.metadata = itemInfo;
+    },
+    async saveMetadata() {
+      const item = pick(this.metadata, [
```
[the remark this answers is not in the document]
Comment (Dec-F) [reply components/MetadataEditor.vue:200]: Because the input parameters and output parameters of this interface design are inconsistent, such as ForcedSortName and SortName. Not sure if there are other problems, so I copied the implementation of jf-web

File: components/MetadataEditor.vue [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        // TODO: show success toast
+        console.log('saved');
+        this.$emit('save');
+      } catch (err) {
+        console.log(err);
+      }
+    },
+    saveDate(key: string, date: string) {
+      this.menu = false;
+      this.metadata = Object.assign({}, this.metadata, {
```
[the remark this answers is not in the document]
Comment (Dec-F) [reply components/MetadataEditor.vue:253]: Uh, my fault, I forgot what I thought

File: components/MetadataEditor.vue [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+      menu: false,
+      dialog: false,
+      person: null as BaseItemPerson | null
+    };
+  },
+  computed: {
+    premiereDate: {
+      get() {
+        if (!this.metadata.PremiereDate) return '';
+        const dateStr = format(
```
[the remark this answers is not in the document]
Comment (Dec-F) [reply components/MetadataEditor.vue:162]: The locale.code of date-fns is inconsistent with that of i18n. For example, it is en in i18n, but en-US in date-dns.

Need to think about how to be consistent

[the remark this answers is not in the document]
Comment (Dec-F) [reply components/MetadataEditor.vue:162]: The time display is generally digital, and there are few scenes that require i18n. I think this can be put on hold

File: components/MetadataEditorDialog.vue [hunk 10]
```diff
@@ -0,0 +1,36 @@
+<template>
+  <v-dialog
+    :value="dialog"
+    max-width="60%"
```
[the remark this answers is not in the document]
Comment (Dec-F) [reply components/MetadataEditorDialog.vue:4]: There are height restrictions inside v-dialog, and the responsive implementation can be implemented by another PR

File: pages/metadata/index.vue (hunk 4 above)
Comment (Dec-F) [reply pages/metadata/index.vue:62]: This interface does not support generics. The format returned by the interface depends on whether the itemId is empty, so there are two solutions:
1. Assert to the correct type, as I do now
2. Modify the type of getItem so that it supports generics.

But if the type file is automatically generated, no matter what kind of scheme, we can't prompt us when the interface changes.

**Comment by dkanada** [comment]
@Camc314 nice, I see you've got a great system for snackbars and dialogs working!

**Comment by KristupasSavickas** [comment]
I get prompted to accept the use of `Google's location service` when I press `Edit Metadata` on a card. What's up with it? Also, nothing happens after I accept/decline.

**Comment by KristupasSavickas** [comment]
I also get this warning, maybe it's related to the dialog not showing up:
```
vue.runtime.esm.js?2b0e:619 [Vue warn]: Avoid mutating a prop directly since the value will be overwritten whenever the parent component re-renders. Instead, use a data or computed property based on the prop's value. Prop being mutated: "dialog"

found in

---> <EditMetadata> at components/EditMetadata.vue
       <ItemMenu> at components/ItemMenu.vue
         <NuxtLink>
           <Card> at components/Card.vue
             <VueperSlide>
               <VueperSlides>
                 <HomeSection> at components/HomeSection.vue
                   <Pages/index.vue> at pages/index.vue
                     <Nuxt>
                       <VMain>
                         <VApp>
                           <Layouts/default.vue> at layouts/default.vue
                             <Root>
```

**Comment by Dec-F** [comment]
> I get prompted to accept the use of `Google's location service` when I press `Edit Metadata` on a card. What's up with it? Also, nothing happens after I accept/decline.

The feature of meta-information editing has not been integrated into the main warehouse, you can go to my personal warehouse to get the latest code experience

**Comment by KristupasSavickas** [comment]
> warehouse

I assume you mean branch? I tested what's in this PR, should I look somewhere else?

File: .prettierrc [hunk 11]
```diff
@@ -1,5 +1,6 @@
 {
   "semi": true,
   "singleQuote": true,
-  "trailingComma": "none"
+  "trailingComma": "none",
+  "htmlWhitespaceSensitivity": "ignore"
```
Comment (Camc314) [inline .prettierrc:5]: Why is this added?

File: components/ItemMenu.vue [hunk 12]
```diff
@@ -0,0 +1,69 @@
+<template>
+  <div>
+    <v-menu absolute>
+      <template v-slot:activator="{ on, attrs }">
+        <v-btn
+          class="card-more-button"
+          icon
+          smal
```
Comment (Camc314) [inline components/ItemMenu.vue:8]:
```suggestion
          small
```

File: components/MetadataEditor.vue [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+      try {
+        this.loading = true;
+        await this.$api.itemUpdate.updateItem({
+          itemId: this.metadata.Id as string,
+          baseItemDto: item
+        });
+        this.$emit('save');
+        this.loading = false;
+        this.$store.dispatch('snackbar/display', {
+          message: 'saved',
```
Comment (Camc314) [inline components/MetadataEditor.vue:332]:
```suggestion
          message: this.$t('saved'),
```

File: locales/en-US.json [hunk 14]
```diff
@@ -34,5 +34,52 @@
...
+  "primary": "Primary",
+  "art": "Art",
+  "backdrop": "Backdrop",
+  "banner": "Banner",
+  "box": "Box",
+  "boxRear": "Box (rear)",
+  "disc": "Disc",
+  "logo": "Logo",
+  "menu": "Menu",
+  "screenshot": "Screenshot",
```
Comment (Camc314) [inline locales/en-US.json:76]:
```suggestion
  "screenshot": "Screenshot",
  "saved": "Saved",
  "cancel": "Cancel",
  "save": "Save",
```

File: components/PersonEditor.vue [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+          <v-text-field
+            v-if="editState.Type === 'Actor'"
+            v-model="editState.Role"
+            outlined
+            :label="$t('role')"
+          ></v-text-field>
+          <v-card-actions>
+            <v-spacer></v-spacer>
+            <v-btn depressed @click="handleCancle">Cancel</v-btn>
+            <v-btn depressed color="primary" type="submit">Save</v-btn>
```
Comment (Camc314) [inline components/PersonEditor.vue:27]:
```suggestion
            <v-btn depressed @click="handleCancle">{{ $t('cancel') }}</v-btn>
            <v-btn depressed color="primary" type="submit">{{ $t('save') }}</v-btn>
```

File: components/MetadataEditor.vue [hunk 16]
```diff
@@ -0,0 +1,407 @@
...
+          </v-tabs-items>
+        </v-col>
+      </v-row>
+    </v-card-text>
+
+    <v-card-actions class="d-flex justify-center align-center">
+      <v-btn @click="$emit('cancel')">Cancel</v-btn>
+      <v-btn color="primary" :loading="loading" @click="saveMetadata">
+        Save
+      </v-btn>
```
Comment (Camc314) [inline components/MetadataEditor.vue:178]:
```suggestion
      <v-btn @click="$emit('cancel')">{{ $t('cancel') }}</v-btn>
      <v-btn color="primary" :loading="loading" @click="saveMetadata">
        {{ $t('save') }}
      </v-btn>

File: components/MetadataEditor.vue [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+  border-bottom: 1px solid var(--v-secondary-lighten1);
+}
+.metadata-editor >>> .v-tab {
+  border-right: 1px solid var(--v-secondary-lighten1);
+}
+.metadata-editor >>> .v-card__actions {
+  position: absolute;
+  bottom: 10px;
+  width: 100%;
+  background-color: #1e1e1e;
```
Comment (Camc314) [inline components/MetadataEditor.vue:392]:
```suggestion
```
This isn't needed as the `v-card` already has a background color. It also causes issues on light mode

File: components/MetadataEditor.vue [hunk 18]
```diff
@@ -0,0 +1,407 @@
...
+}
+.headline {
+  border-bottom: 1px solid var(--v-secondary-lighten1);
+}
+.metadata-editor >>> .v-tab {
+  border-right: 1px solid var(--v-secondary-lighten1);
+}
+.metadata-editor >>> .v-card__actions {
+  position: absolute;
+  bottom: 10px;
```
Comment (Camc314) [inline components/MetadataEditor.vue:390]:
```suggestion
```
This is causing issues for the action icons in the image editor to overlay onto the text

File: components/MetadataEditorDialog.vue [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    itemId: {
+      type: String,
+      default: ''
+    }
+  },
+  data() {
+    return {
+      forceRefresh: false
+    };
+  },
```
Comment (Camc314) [inline components/MetadataEditorDialog.vue:36]:
```suggestion
  },
  computed: {
    calcMaxWidth() {
      return window.innerWidth < 700 ? '100%' : '60%';
    }
  },
```
This makes the metadata slightly more friendly on smaller screens.

File: components/ItemMenu.vue [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+        return [
+          {
+            title: this.$t('editMetadata') as string,
+            action: () => {
+              this.dialog = true;
+            }
+          }
+        ];
+      }
+    }
```
Comment (Camc314) [inline components/ItemMenu.vue:58]: We probably need to check here is the user is an admin before letting them to attempt to edit metadata.
The property is accessible via:
```this.$auth.user.Policy.IsAdministrator```

**Review by Camc314 — COMMENTED** [review]
Sorry for taking so long to review.

This is just some initial thoughts I have noticed

File: components/MetadataEditorDialog.vue (hunk 19 above)
Comment (MrTimscampi) [reply components/MetadataEditorDialog.vue:36]: We should use the Vuetify breakpoints for this.

See https://vuetifyjs.com/en/features/breakpoints/#breakpoint-service-object

Comment (MrTimscampi) [reply components/MetadataEditorDialog.vue:36]: Also for mobile, the dialog should likely be full screen, with the tabs moved to an horizontal tab bar at the top.

Comment (Camc314) [reply components/MetadataEditorDialog.vue:36]: Yep that is a better method

**Comment by MrTimscampi** [comment]
To add to Cameron's review (this isn't a formal technical review, but more of a functional one), I've had a few bugs pop up while testing this:

When closing a modal by clicking outside of it, it reopens when switching tabs:
Screenshot_2020-11-02 jellyfin-vue(2)

Resolution is sometimes cut off:
Screenshot_2020-11-02 jellyfin-vue(1)

There is no way to add images, only to delete them or change the existing ones:
Screenshot_2020-11-02 jellyfin-vue

The forms are cutoff at the bottom:
Screenshot_2020-11-02 jellyfin-vue(3)

**Comment by Camc314** [comment]
Rebased

File: components/ItemMenu.vue [hunk 21]
```diff
@@ -0,0 +1,71 @@
+<template>
+  <div>
+    <v-menu absolute>
+      <template v-slot:activator="{ on, attrs }">
+        <v-btn
```
Comment (Camc314) [inline components/ItemMenu.vue:5]: This is causing an ESLint error. However changing it to `#:activator` breaks it :P

**Comment by Camc314** [comment]
@MrTimscampi How much more should we do on this PR? There's a few things that still need addressing, but I think it might be easier to merge, and then address them later. All the core functionality is there

**Comment by MrTimscampi** [comment]
> @MrTimscampi How much more should we do on this PR? There's a few things that still need addressing, but I think it might be easier to merge, and then address them later. All the core functionality is there

I have a bunch of fixes and dialog design cleanup locally. Finishing the last modal and then I'll push. After that I think it's good to go.

**Comment by MrTimscampi** [comment]
@Camc314 My fixes are pushed. There's a few stuff I didn't manage to fix (Overflow on the image search, most notably) and at least two of the dialogs still need a responsive pass.

I've fiddled with this all afternoon and evening though, so I'm kind of done with it for now :smile:

File: nuxt.config.ts [hunk 22]
```diff
@@ -159,6 +160,10 @@ const config: NuxtConfig = {
       fallbackLocale: 'en'
     }
   },
+  dateFns: {
+    locales: ['en-US', 'zh-CN'],
+    defaultLocale: 'en-US'
+  },
```
Comment (Camc314) [inline nuxt.config.ts:166]: Not quite sure why there are only two locales here

File: components/ImageEditor.vue [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+      </v-col>
+    </v-row>
+    <v-row v-if="backdropImages.length">
+      <v-col class="pt-0">
+        <h2 class="text-h6">{{ $t('backdrop') }}</h2>
+        <div>
+          <v-card
+            v-for="(item, i) in backdropImages"
+            :key="i"
+            class="mx-2"
```
Comment (Camc314) [inline components/ImageEditor.vue:43]:
```suggestion
            :key="`${item.ImageTag}-${i}`"

File: components/ImageSearch.vue [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+          :disabled="loading"
+          hide-details
+        ></v-checkbox>
+      </v-row>
+      <v-divider />
+      <v-row class="image-results">
+        <v-col class="card-grid-container">
+          <v-card
+            v-for="(item, i) in images"
+            :key="i"
```
Comment (Camc314) [inline components/ImageSearch.vue:44]:
```suggestion
            :key="`${item.Type}-${i}`"

File: components/PersonEditor.vue [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+        :class="{
+          'justify-end': !$vuetify.breakpoint.mobile,
+          'justify-center': $vuetify.breakpoint.mobile
+        }"
+      >
+        <v-spacer></v-spacer>
+        <v-btn depressed width="8em" class="mr-1" @click="handleCancel">
+          Cancel
+        </v-btn>
+        <v-btn depressed width="8em" color="primary" type="submit">Save</v-btn>
```
Comment (Camc314) [inline components/PersonEditor.vue:52]: need to be translated

File: components/MetadataEditor.vue [hunk 26]
```diff
@@ -0,0 +1,429 @@
...
+              <template #no-data>
+                <v-list-item>
+                  <v-list-item-content>
+                    <v-list-item-title>
+                      No results matching "
+                      <strong>{{ search }}</strong>
+                      ". Press
+                      <kbd>enter</kbd>
+                      to create a new one
+                    </v-list-item-title>
```
Comment (Camc314) [inline components/MetadataEditor.vue:109]: Translateddd

File: components/MetadataEditor.vue [hunk 27]
```diff
@@ -0,0 +1,429 @@
...
+                  </v-list-item-content>
+                </v-list-item>
+              </template>
+            </v-combobox>
+          </v-tab-item>
+          <v-tab-item value="castAndCrew">
+            <v-list two-line>
+              <v-list-item @click="(e) => handlePersonEdit()">
+                <v-list-item-content>
+                  <v-list-item-title>Add a new person</v-list-item-title>
```
Comment (Camc314) [inline components/MetadataEditor.vue:143]: Translated

**Review by Camc314 — COMMENTED** [review]
Few comments, mainly regarding translations, and v-for keys

**Comment by Camc314** [comment]
> @Camc314 My fixes are pushed. There's a few stuff I didn't manage to fix (Overflow on the image search, most notably) and at least two of the dialogs still need a responsive pass.
> 
> I've fiddled with this all afternoon and evening though, so I'm kind of done with it for now 😄

Cool, fair enough. I've just added translations for some we missed, and updated some of the v-for loops to use more unique keys (`${itemId}-${index}`)

I've left once comment about the `date-fns` config in `nuxt.config.ts`

File: components/ImageEditor.vue [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+            v-for="(item, i) in generalImages"
+            :key="`${item.ImageTag}-${i}`"
+            class="ma-2"
+            outlined
+            width="288"
+          >
+            <v-img
+              :src="imageFormat(item)"
+              contain
+              width="288"
```
Comment (ferferga) [inline components/ImageEditor.vue:17]: Why this fixed size

File: pages/metadata/index.vue [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+
+      (node.children as ITreeNode[]).push(
```
Comment (ferferga) [inline pages/metadata/index.vue:63]: Refactor

File: mixins/localeHelper.ts [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+
+      if (i18nlocale === 'en') {
+        locale = 'enUS';
+      } else if (i18nlocale === 'zh') {
+        locale = 'zhCN';
+      } else {
+        locale = i18nlocale;
+      }
+      return locale;
+    }
```
[the remark this answers is not in the document]
Comment (MrTimscampi) [reply mixins/localeHelper.ts:38]: These have special codes in date-fns, the others are just the standard two letter codes, I believe.

[the remark this answers is not in the document]
Comment (Camc314) [reply mixins/localeHelper.ts:38]: Ah Ok, that makes sense. Is there anything left to be done or can this be merged?

**Review by MrTimscampi — APPROVED** [review]

**Comment by MrTimscampi** [comment]
Let's merge this as-is and fix it up later on.
